<a href="https://colab.research.google.com/github/azizbekb/NumPy/blob/main/amaliyot_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Dastlab, kerakli ma'lumotlarni pastki qatordagi cellni yuritish orqali yuklab olamiz:

In [ ]:
!wget -O GM_Cars.zip "https://drive.google.com/uc?export=download&id=1i57swZxVA9Xh64eVvylnkIJmKT-4WTZy" && unzip GM_Cars.zip && rm -r GM_Cars.zip

In [1]:
# Numpy kutubxonasini chaqirib olish
import numpy as np

### Masala(1): Ma'lumotlarni yuklab olish.

**Ma'lumotlar :** Ma'lumotlar ikkita faylda joylangan bo'lib, bu fayllar nomlari `gm_2020.npz` va `gm_2021.npz` ko'rinishda, bu fayllar mos ravishda GM Uzbekistan tomonidan 2020 va 2021 yillarning yanvar-iyul oylarida ishlab chiqarilgan avtomobillar modellari va ularning sonlari keltirilgan. `npz` kengaytmasidan ko'rinib turibdiki, xar bir fayl o'z ichiga massivlarni(bizning holatda ikkita massiv, (1) Avtomobil modellari nomi, (2) Ishlab chiqarilgan avtomobillar soni) mujassam etgan. 

Masala sharti shundan iboratki, har bir faylni yuklab ularni ikkita mos massivga ajratuvchi funksiyani davom ettiring. Tuzilajak funksiyangiz argument sifatida fayl nomini qabul qilishini ta'minlang. 

**Muhim**: Massivlar: `models` va `qnts` nomi ostida saqlangan.

**Yordam :**  `load` metodi yordamida yuklab, so'ngra massiv nomlaridan foydalanib natijani qaytaring. 

**Natija**: Agar funksiyaga `gm_2020.npz` faylini uzatilsa, natija quyidagi ko'rinishda bo'lishi mumkin:

(array(['Labo', 'Damas', ......., 'Gentra'], dtype='<U6'), array([ 4521, 30889, ......,  30368]))



***Ma'lumot manbasi***: *gazeta.uz*

In [8]:
# JAVOBNI SHU YERGA YOZING
def load_data(file_name):
  """Har bir faylni yuklab ularni ikkita mos massivga ajratuvchi funksiyani tuzish """
  data=np.load(file_name)
  models=data['models']
  qnts=data['qnts']
  return models, qnts
print(load_data('GM_Cars/gm_2020.npz'))
print(load_data('GM_Cars/gm_2021.npz'))

(array(['Labo', 'Damas', 'Spark', 'Nexia', 'Cobalt', 'Gentra'], dtype='<U6'), array([ 4521, 30889, 16370, 41663, 44901, 30368], dtype=int32))
(array(['Labo', 'Damas', 'Spark', 'Nexia', 'Cobalt', 'Gentra'], dtype='<U6'), array([ 7542, 33682,  6485, 14442, 36317, 26266], dtype=int32))


### Masala(2): Jami ishlab chiqarilgan avtomobillar soni.

Yuqorida yaratilingan ma'lumotlarni yuklab ajratib oluvchi funksiyamizdan foydalanib, endilikda ma'lumotlar ustida amallar bajaramiz. 

**Masala :** Ma'lumotlardan foydalanib, 2020 va 2021  yillarda ishlab chiqarilgan jami avtomobillar sonini hisoblaydigan funksiyani davom ettiring.

**Natija**: (168..., ...734) ko'rinishida bo'ladi. 
 
Bu yerda 168.... ta avtomabil 2020 yilda ishlab chiqarilgan. 2021 yilda esa ...734 ta avtomabil ishlab chiqarilan. 

In [9]:
# JAVOBNI SHU YERGA YOZING
def total_qnts():
    """Ma'lumotlardan foydalanib, 2020 va 2021 yillarda ishlab chiqarilgan jami avtomobillar sonini hisoblaydigan 
    funksiya yaratish."""
    data1 = load_data('GM_Cars/gm_2020.npz') # ma'lumotlar (2020)
    data2 = load_data('GM_Cars/gm_2021.npz') # ma'lumotlar (2021)
    total_qnts_2020=np.sum(data1[1])
    total_qnts_2021=np.sum(data2[1])
    return total_qnts_2020, total_qnts_2021
print(total_qnts())


(168712, 124734)


### Masala(3): Ba'zi avtomabillarning ishlab chiqarilish sonlari. 



**Masala :** Tasavvur qiling, ish beruvchi sizdan faqatgina `Damas`, `Nexia`, `Cobalt`, va `Gentra`rusumli avtomabillarni 2020 va 2021 yillarda qancha ishlab chiqarilganini barcha ma'lumotlardan ajratib berishni so'radi. Siz, shu vazifani bajaradigan funksiyani davom ettirishingiz kerak. 

 **Yordam**: Boolean indekslash mavzusiga murojat qiling.
 
 
 **Natija**: (array([30889, .....  30368]), array([33682 .....  26266])) ko'rinishida bo'ladi

In [10]:
# JAVOBNI SHU YERGA YOZING
def sel_models():
    """Faqatgina Damas, Nexia, Cobalt, va Gentra rusumli avtomabillarni 2020 va 2021 yillarda qancha ishlab chiqarilganini 
    barcha ma'lumotlardan ajratib olish"""
    data1 = load_data('GM_Cars/gm_2020.npz') # ma'lumotlar (2020)
    data2 = load_data('GM_Cars/gm_2021.npz') # ma'lumotlar (2021)
    musk1 = (data1[0] == 'Damas') | (data1[0] == 'Nexia') | (data1[0] == 'Cobalt') | (data1[0] == 'Gentra')
    x = data1[1][musk1]
    musk2 = (data2[0] == 'Damas') | (data2[0] == 'Nexia') | (data2[0] == 'Cobalt') | (data2[0] == 'Gentra')
    y = data2[1][musk2]
    return x, y
print(sel_models())

(array([30889, 41663, 44901, 30368], dtype=int32), array([33682, 14442, 36317, 26266], dtype=int32))


### Masala(4): Ba'zi avtomabillarning ishlab chiqarilish sonlari umumiy ishlab chiqarilgan avtomabilarning qancha foizini tashkil etadi. 


Yuqorida faqatgina `Damas`, `Nexia`, `Cobalt`, va `Gentra`rusumli avtomabillarni 2020 va 2021 yillarda qancha ishlab chiqarilganini barcha ma'lumotlardan ajratib oldik. 


**Masala :** Ana endilikda shu avtomabillarni jami ishlab chiqarilgan avtomabillarning qancha foizini tashkil etishini hisoblab ko'ramiz. Shu vazifani bajaruvchi funksiyani tuzing. 

 **Yordam**: Masala(2)da jami avtomabillarning sonini hisoblab chiqgan edik, Masala(3)da esa ba'zi avtomabillarning sonini hisoblab topdik. Shu ma'lumotlardan foydalaning. 
 
 **Hisoblash formulasi:** $$\frac{Avtomabil\;soni}{Jami\; avtomabillar\; soni}*100% $$
 
 
 **Natija**: (array([18.30871544, ..... 17.99990516]),
 array([27.00306252, ..... 21.0576106 ])) ko'rinishida bo'ladi
 
 Ya'ni, `Damas` rusumli avtomabil 2020 yilda jami ishlab chiqarishning 18.3% ni tashkil etadi 2021 yilda esa 27% ni tashkil etmoqda va h.k.
 

In [11]:
# JAVOBNI SHU YERGA YOZING
def ratio_sel_models():
    """Ba'zi avtomabillarning ishlab chiqarish hajmi, jami ishlab chiqarilgan avtomabillar hajmining qanday 
    foizini egallashini hisoblash funksiyasi"""
    data1=total_qnts()
    data2=sel_models()
    result1=(data2[0]/data1[0])*100
    result2=(data2[1]/data1[1])*100
    return result1,result2
print(ratio_sel_models())

(array([18.30871544, 24.69474608, 26.61399308, 17.99990516]), array([27.00306252, 11.57823849, 29.11555791, 21.0576106 ]))
